packages/server/src/sync/api.ts

```diff
@@ -1,38 +1,65 @@
+import { SyncChannels, SyncRequest, SyncWebhook } from '@botpress/messaging-base'
 import { Response } from 'express'
 import Joi from 'joi'
+import _ from 'lodash'
 import { ApiManager } from '../base/api-manager'
 import { ClientApiRequest } from '../base/auth/client'
 import { ChannelService } from '../channels/service'
 import { makeSyncRequestSchema } from './schema'
 import { SyncService } from './service'
 
-export class SyncApi {
-  private schema!: Joi.ObjectSchema
+const LEGACY_VERSION = '0.1.0'
 
+export class SyncApi {
   constructor(private syncs: SyncService, private channels: ChannelService) {}
 
   setup(router: ApiManager) {
-    this.schema = makeSyncRequestSchema(this.channels.list())
-    router.post('/sync', this.schema, this.sync.bind(this))
+    router.post('/sync', makeSyncRequestSchema(this.channels.list()), this.sync.bind(this))
   }
 
   async sync(req: ClientApiRequest, res: Response) {
-    const body = { ...req.body, channels: this.removeDisabledChannels(req.body.channels || {}) }
-    const { value } = this.schema.validate(body)
+    const { value, error } = this.validate(req.body)
+    if (error) {
+      return res.status(400).send(error.message)
+    }
 
-    const result = await this.syncs.sync(req.clientId, value)
+    const result = await this.syncs.sync(req.clientId, value!)
     res.send(result)
   }
 
-  private removeDisabledChannels(body: any) {
-    const filtered: any = {}
+  validate(req: SyncRequest): { value?: SyncRequest; error?: Joi.ValidationError } {
+    const webhooks: SyncWebhook[] = req.webhooks || []
+    const channels: SyncChannels = req.channels || {}
+
+    for (const [name, config] of Object.entries(channels)) {
+      const channel = this.channels.getByNameAndVersion(name, config?.version || LEGACY_VERSION)
+
+      const { error, value } = Joi.object({
+        body: {
+          channels: {
+            [channel.meta.name]: {
+              version:
+                channel.meta.version === LEGACY_VERSION
+                  ? Joi.string().valid(channel.meta.version).optional()
+                  : Joi.string().valid(channel.meta.version).required(),
+              enabled: Joi.boolean().optional(),
+              ...channel.meta.schema
+            }
+          }
+        }
+      })
+        .options({ stripUnknown: channel.meta.version === LEGACY_VERSION, abortEarly: false })
+        .validate({ body: { channels: { [channel.meta.name]: config } } })
```

Comment:
We only stripUknown for legacy channels. We will enforce correct payload formats for new channels. Here we're basically doing custom validation, but since we want the error messages to be consistent, we put the object to be validated inside body.channels so that all error message look like this : "error somewhere in body.channels.channelName etc"